### PR TITLE
Add Session Id blinding

### PR DESCRIPTION
**Pull request by ceokot**
Including modified version of lazysodium-android to expose missing libsodium functions, we could build from a fork which we still need to setup.

### CODE REVIEW

File: libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt [hunk 1]
```diff
@@ -530,7 +530,7 @@ object SnodeAPI {
             400, 500, 502, 503 -> { // Usually indicates that the snode isn't up to date
                 handleBadSnode()
             }
-            406 -> {
+            425 -> {
```
Comment (hjubb) [inline libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt:533]: I wonder if we will need to maintain both of these for this class of errors

**Review by hjubb — COMMENTED** [review]
Just some general initial stuff in case that open group signature wasn't verifying correctly, otherwise I'm liking how it all looks and the layout / formatting of where everything lives is really nice and easy to read through :tada:

File: libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt (hunk 1 above)
Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt:533]: Yes, going to revert this bit.

File: libsession/src/main/java/org/session/libsession/messaging/open_groups/OpenGroupApiV4.kt [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+        if (request.verb == GET) {
+            for ((key, value) in request.queryParameters) {
+                urlBuilder.addQueryParameter(key, value)
+            }
+        }
+        fun execute(token: String?): Promise<Map<*, *>, Exception> {
+            val publicKey =
+                MessagingModuleConfiguration.shared.storage.getOpenGroupPublicKey(request.server)
+                    ?: return Promise.ofFail(Error.NoPublicKey)
+            val ed25519KeyPair = MessagingModuleConfiguration.shared.getUserED25519KeyPair()
```
[the remark this answers is not in the document]
Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/messaging/open_groups/OpenGroupApiV4.kt:143]: I'll clean this up

File: libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt [hunk 3]
```diff
@@ -3,27 +3,31 @@ package org.session.libsession.messaging.utilities
...
 import com.goterl.lazysodium.utils.KeyPair
 import org.session.libsignal.utilities.Hex
 import org.session.libsignal.utilities.toHexString
 import kotlin.experimental.inv
 
 object SodiumUtilities {
-
     private val sodium by lazy { LazySodiumAndroid(SodiumAndroid()) }
 
+    private const val scalarLength: Int = 32 // crypto_core_ed25519_scalarbytes
```
Comment (hjubb) [inline libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt:15]: should we standardise constant values to all caps with underscores? probably a decision that we could discuss and do consistently across new code

Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt:15]: Certainly, I'll update these, thanks

File: app/src/main/java/org/thoughtcrime/securesms/conversation/v2/messages/VisibleMessageView.kt [hunk 4]
```diff
@@ -123,7 +127,14 @@ class VisibleMessageView : LinearLayout {
             binding.profilePictureView.glide = glide
             binding.profilePictureView.update(message.individualRecipient)
             binding.profilePictureView.setOnClickListener {
-                showUserDetails(senderSessionID, threadID)
+                if (thread.isOpenGroupRecipient && IdPrefix.fromValue(senderSessionID) == IdPrefix.BLINDED) {
```
Comment (hjubb) [inline app/src/main/java/org/thoughtcrime/securesms/conversation/v2/messages/VisibleMessageView.kt:130]: Should we move this check into the `UserDetailsBottomSheet` so the user isn't just presented with a new activity for blinded IDs?

Comment (ceokot) [reply app/src/main/java/org/thoughtcrime/securesms/conversation/v2/messages/VisibleMessageView.kt:130]: Yes, will remove this

**Review by hjubb — COMMENTED** [review]
I think I'll have to look over some higher level documentation to check if some of the blinding functions but like the overall improvements for code quality across the parts we probably don't improve enough over new features :)

File: libsession/src/main/java/org/session/libsession/messaging/file_server/FileServerApi.kt [hunk 5]
```diff
@@ -73,29 +72,37 @@ object FileServerAPIV2 {
...
         }
     }
 
     fun upload(file: ByteArray): Promise<Long, Exception> {
         val base64EncodedFile = Base64.encodeBytes(file)
         val parameters = mapOf( "file" to base64EncodedFile )
-        val request = Request(verb = HTTP.Verb.POST, endpoint = "files", parameters = parameters)
-        return send(request).map { json ->
-            json["result"] as? Long ?: throw OpenGroupAPIV2.Error.ParsingFailed
+        val request = Request(
```
[the remark this answers is not in the document]
Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/messaging/file_server/FileServerApi.kt:89]: Yes this has changed as seen here.

File: libsession/src/main/java/org/session/libsession/messaging/open_groups/OpenGroupApi.kt [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+    private val timeSinceLastOpen by lazy {
+        val context = MessagingModuleConfiguration.shared.context
+        val lastOpenDate = TextSecurePreferences.getLastOpenTimeDate(context)
+        val now = System.currentTimeMillis()
+        now - lastOpenDate
+    }
+
+    private const val defaultServerPublicKey =
+        "a03c383cf63c3c4efe67acc52112a6dd734b3a946b9545f488aaa93da7991238"
+    const val defaultServer = "http://116.203.70.33"
```
[the remark this answers is not in the document]
Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/messaging/open_groups/OpenGroupApi.kt:62]: Yeah that sounds good, happy to include that update in this PR

File: libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        // k = salt.crypto_core_ed25519_scalar_reduce(blake2b(server_pk, digest_size=64).digest())
+        val serverPubKeyData = Hex.fromStringCondensed(serverPublicKey)
+        val serverPubKeyHash = ByteArray(GenericHash.BLAKE2B_BYTES_MAX)
+        if (!sodium.cryptoGenericHash(serverPubKeyHash, serverPubKeyHash.size, serverPubKeyData, serverPubKeyData.size.toLong())) {
+            return null
+        }
+        // Reduce the server public key into an ed25519 scalar (`k`)
+        val x25519PublicKey = ByteArray(SCALAR_LENGTH)
+        sodium.cryptoCoreEd25519ScalarReduce(x25519PublicKey, serverPubKeyHash)
+        return if (x25519PublicKey.any { it.toInt() != 0 }) {
```
[the remark this answers is not in the document]
Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt:36]: We are checking that`x25519PublicKey` that was previously intialized with `0` was updated by the function call so an alternative to the above check might look like below, using `.any` was just a personal preference:
`!x25519PublicKey.all { it.toInt() == 0 }`

File: app/src/main/java/org/thoughtcrime/securesms/groups/OpenGroupUtilities.kt [hunk 8]
```diff
@@ -29,8 +29,7 @@ object OpenGroupUtilities {
             throw IllegalStateException("Attempt to update open group info for non-existent DB record: $groupId")
         }
 
-        val info = OpenGroupAPIV2.getInfo(room, server).get() // store info again?
-        OpenGroupAPIV2.getMemberCount(room, server).get()
+        val info = OpenGroupApi.getRoomInfo(room, server).get() // store info again?
```
[the remark this answers is not in the document]
Comment (ceokot) [reply app/src/main/java/org/thoughtcrime/securesms/groups/OpenGroupUtilities.kt:32]: We do in OpenGroupPoller

File: libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (hjubb) [reply libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt:36]: that makes sense, I was just wondering the case where if this fails all of them would be zero or some weird hypothetical case where for some reason all of them except one was non-zero but it makes sense to me now :+1: does the `cryptoCoreEd25519ScalarReduce` return a boolean we can use if it's a failure or I guess the resulting byte array would be all zeroes

[the remark this answers is not in the document]
Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/messaging/utilities/SodiumUtilities.kt:36]: The function has a void return type.

**Review by hjubb — COMMENTED** [review]
Looks good :+1: 
There is a new migration in the SQLCipherOpenHelper from the hot fix build you might have to merge in. but it should just be the number ordering

edit, nevermind looks like you already fixed those merge conflicts

**Review by hjubb — APPROVED** [review]

File: libsession/src/main/java/org/session/libsession/messaging/sending_receiving/pollers/OpenGroupPoller.kt [hunk 9]
```diff
@@ -139,7 +139,7 @@ class OpenGroupPoller(private val server: String, private val executorService: S
             OpenGroupMessage(
                 serverID = it.id,
                 sender = it.sessionId,
-                sentTimestamp = TimeUnit.SECONDS.toMillis(it.posted),
+                sentTimestamp = TimeUnit.SECONDS.toMillis((it.posted * 1000).toLong()),
```
Comment (hjubb) [inline libsession/src/main/java/org/session/libsession/messaging/sending_receiving/pollers/OpenGroupPoller.kt]: Does the `TimeUnit.SECONDS.toMillis` need to still be here after the * 1000? I'm not sure what `OpenGroupMessage.fromJson()` was doing

Comment (ceokot) [reply libsession/src/main/java/org/session/libsession/messaging/sending_receiving/pollers/OpenGroupPoller.kt]: That's not meant to be there, thanks. I picked this up in my testing earlier and fixed it in my last commit.

**Pull request merged**